api/src/auth/drivers/saml.ts

```diff
@@ -0,0 +1,172 @@
+			.select('id')
+			.from('directus_users')
+			.whereRaw('LOWER(??) = ?', ['external_identifier', identifier.toLowerCase()])
+			.first();
+
+		return user?.id;
+	}
+
+	async getUserID(payload: Record<string, any>) {
+		const { provider, emailKey, identifierKey, nameKey, givenNameKey, familyNameKey, allowPublicRegistration } =
+			this.config;
+
+		const email = payload[emailKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress'];
+		const identifier = payload[identifierKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier'];
+		const userID = await this.fetchUserID(identifier);
+
+		const userInfo = {
+			name: payload[nameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'],
+			given_name: payload[givenNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'],
+			family_name: payload[familyNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'],
+			email,
+		};
+
+		if (userID) return userID;
+
+		// Is public registration allowed?
+		if (!allowPublicRegistration) {
+			logger.trace(`[SAML] User doesn't exist, and public registration not allowed for provider "${provider}"`);
+			throw new InvalidCredentialsException();
+		}
+
+		try {
+			await this.usersService.createOne({
+				provider,
+				first_name: userInfo.given_name,
+				last_name: userInfo.family_name,
+				email: email,
+				external_identifier: identifier.toLowerCase(),
+				role: this.config.defaultRoleId,
+			});
+			const userID = await this.fetchUserID(identifier);
+			if (userID) return userID;
+		} catch (e) {
+			if (e instanceof RecordNotUniqueException) {
+				logger.warn(e, '[SAML] Failed to register user. User not unique');
+				throw new InvalidProviderException();
+			}
+			throw e;
+		}
+
+		throw new InvalidCredentialsException();
+	}
+
+	async login(user: User): Promise<void> {
+		if (!user.auth_data) {
+			throw new InvalidCredentialsException();
+		}
+	}
+}
+
+export function createSAMLAuthRouter(providerName: string) {
+	const router = Router();
+
+	router.get(
+		'/metadata',
+		asyncHandler(async (req, res) => {
+			const { sp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			return res.header('Content-Type', 'text/xml').send(sp.getMetadata());
+		})
+	);
+
+	router.get(
+		'/',
+		asyncHandler(async (req, res) => {
+			const { sp, idp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			const { context: url } = await sp.createLoginRequest(idp, 'redirect');
+			const parsedUrl = new URL(url);
+			if (req.query.redirect) {
+				parsedUrl.searchParams.append('RelayState', req.query.redirect as string);
+			}
+			return res.redirect(parsedUrl.toString());
+		})
+	);
+
+	router.post(
+		'/logout',
+		asyncHandler(async (req, res) => {
+			const { sp, idp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			const { context } = await sp.createLogoutRequest(idp, 'redirect', req.body);
+			return res.redirect(context);
+		})
+	);
```

Comment:
**question:** Is this intended to be user initiated, or part of the Directus logout flow? Currently calling this directly won't log you our of Directus (unless I'm mistaken).

Reply:
this is part of federated logout, you set it in your saml config, for example:

```
{
  "logout": {
    "callback": "http://localhost:8055/auth/login/saml/logout",
    "slo_enabled": true
  }
}
```

which is then triggered when you logout, not really a Directus function so much as a saml function.

Reply:
Should we not also invalidate the Directus session when this is called though?

Reply:
yeah that now happens